src/v/redpanda/admin_server.cc

```diff
@@ -1748,3 +1753,73 @@ void admin_server::register_hbadger_routes() {
               [] { return ss::json::json_return_type(ss::json::json_void()); });
       });
 }
+
+void admin_server::register_transaction_routes() {
+    if (!config::shard_local_cfg().enable_transactions) {
+        vlog(logger.info, "Transactions are disabled");
+        return;
+    }
+
+    ss::httpd::transaction_json::get_all_transactions.set(
+      _server._routes,
+      [this](std::unique_ptr<ss::httpd::request> req)
+        -> ss::future<ss::json::json_return_type> {
+          if (need_redirect_to_leader(model::tx_manager_ntp, _metadata_cache)) {
+              throw co_await redirect_to_leader(*req, model::tx_manager_ntp);
+          }
+
+          auto& tx_frontend = _partition_manager.local().get_tx_frontend();
+          if (!tx_frontend.local_is_initialized()) {
+              throw ss::httpd::bad_request_exception(
+                fmt::format("Transaction are disabled"));
+          }
+
+          auto res = co_await tx_frontend.local().get_all_transactions();
+          if (!res.has_value()) {
+              co_await throw_on_error(*req, res.error(), model::tx_manager_ntp);
+          }
+
+          using tx_info = ss::httpd::transaction_json::transaction_summary;
+          std::vector<tx_info> ans;
+          ans.reserve(res.value().size());
+
+          for (auto& [_, tx] : res.value()) {
+              tx_info new_tx;
+
+              new_tx.transactional_id = tx.id;
+
+              ss::httpd::transaction_json::producer_identity pid;
+              pid.id = tx.pid.id;
+              pid.epoch = tx.pid.epoch;
+              new_tx.pid = pid;
+
+              new_tx.tx_seq = tx.tx_seq;
+              new_tx.etag = tx.etag;
+              new_tx.status = tx.get_status();
+              new_tx.timeout_ms = tx.get_timeout().count();
+              new_tx.staleness_ms = tx.get_staleness().count();
+
+              for (auto& partition : tx.partitions) {
+                  ss::httpd::transaction_json::partition partition_info;
+                  partition_info.ns = partition.ntp.ns;
+                  partition_info.topic = partition.ntp.tp.topic;
+                  partition_info.partition_id = partition.ntp.tp.partition;
+                  partition_info.etag = partition.etag;
+
+                  new_tx.partitions.push(partition_info);
+              }
+
+              for (auto& group : tx.groups) {
+                  ss::httpd::transaction_json::group group_info;
+                  group_info.group_id = group.group_id;
+                  group_info.etag = group.etag;
+
+                  new_tx.groups.push(group_info);
+              }
```

Comment:
pre-reserve memory for the vectors and avoid string copies.

Reply:
but it json_list without reserve method

Reply:
sounds good. I blindly assumed these were all vectors!